### PR TITLE
Make `TyKind: Debug` have less verbose output

### PR DESCRIPTION
**Pull request by BoxyUwU**
Current `TyKind: Debug` impl is basically unusable for debugging, its too verbose even for verbose debugging :rofl: This PR replaces the debug logic for `TyKind` with a more manual debug impl instead of a hand expanded derived impl. This should help make #107084 more reasonable to land since the output of `Ty: Debug` will be better.

This isn't a fully completed change to the `Debug` impl of `TyKind` as there's still logic from the derive macro for some variants. Some of the variants are also not consisten with the `-Zverbose` printing of `Ty`, ideally `-Zverbose` printing of `Ty` would also just defer to the debug impl instead of having lots of checks in pretty printing. I plan on fixing this in follow up PRs since it seems tricky to do in this one and its already a large PR :sweat_smile:

### CODE REVIEW

File: compiler/rustc_codegen_cranelift/src/num.rs [hunk 1]
```diff
@@ -368,7 +369,7 @@ pub(crate) fn codegen_ptr_binop<'tcx>(
         .layout()
         .ty
         .builtin_deref(true)
-        .map(|TypeAndMut { ty, mutbl: _ }| !has_ptr_meta(fx.tcx, ty))
+        .map(|RawPtr { ty, mutbl: _ }| !has_ptr_meta(fx.tcx, ty))
```
Comment (compiler-errors) [inline compiler/rustc_codegen_cranelift/src/num.rs:372]: This is misleading. Doing a builtin deref doesn't give you a raw pointer, it gives you the type and mutability of the pointee.

File: compiler/rustc_hir_analysis/src/variance/constraints.rs [hunk 2]
```diff
@@ -226,7 +226,7 @@ impl<'a, 'tcx> ConstraintContext<'a, 'tcx> {
 
             ty::Ref(region, ty, mutbl) => {
                 self.add_constraints_from_region(current, region, variance);
-                self.add_constraints_from_mt(current, &ty::TypeAndMut { ty, mutbl }, variance);
+                self.add_constraints_from_mt(current, &ty::RawPtr { ty, mutbl }, variance);
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/variance/constraints.rs:229]: I feel like this is misleading too

File: compiler/rustc_middle/src/ty/structural_impls.rs [hunk 3]
```diff
@@ -88,7 +88,33 @@ impl fmt::Debug for ty::FreeRegion {
...
-        write!(f, "({:?}; c_variadic: {})->{:?}", self.inputs(), self.c_variadic, self.output())
+        let ty::FnSig { inputs_and_output: _, c_variadic, unsafety, abi } = self;
+
+        write!(f, "{}", unsafety.prefix_str())?;
+        match abi {
+            rustc_target::spec::abi::Abi::Rust => (),
+            abi => write!(f, "extern \"{abi:?}\" ")?,
+        };
+        if *c_variadic {
+            write!(f, "c_variadic ")?;
```
Comment (bjorn3) [inline compiler/rustc_middle/src/ty/structural_impls.rs:99]: Can you instead print `...` as last argument in this case?

File: compiler/rustc_type_ir/src/sty.rs [hunk 4]
```diff
@@ -499,33 +499,53 @@ impl<I: Interner> hash::Hash for TyKind<I> {
 impl<I: Interner> fmt::Debug for TyKind<I> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         match self {
-            Bool => f.write_str("Bool"),
```
Comment (bjorn3) [inline compiler/rustc_type_ir/src/sty.rs:502]: Why this change? If you want a rust like syntax you can print `Ty` instead of `TyKind`. On multiple occasions I have printed `ty.kind()` precisely because I wanted to know the internal representation with all details rather than a nicely formatted representation.

File: compiler/rustc_type_ir/src/sty.rs [hunk 5]
```diff
@@ -499,33 +499,53 @@ impl<I: Interner> hash::Hash for TyKind<I> {
...
+
+                match iter.next() {
+                    None => return write!(f, ")"),
+                    Some(ty) => write!(f, "{ty:?}")?,
+                };
+
+                for ty in iter {
+                    write!(f, ", {:?}", ty)?;
+                }
+                write!(f, ")")
```
Comment (compiler-errors) [inline compiler/rustc_type_ir/src/sty.rs:541]: I don't think this logic is right. It prints `(Ty)` instead of `(Ty,)`

File: compiler/rustc_type_ir/src/sty.rs (hunk 4 above)
Comment (compiler-errors) [reply compiler/rustc_type_ir/src/sty.rs:502]: I don't oppose having a nice syntax as long as it's like... "injective" in the mathematical sense. Like, pretty printing does things like rebinding late bound vars and eliding lifetimes and leaving out defaulted adt substs (I think), but my understanding is that the aim of this PR is to continue conveying the same amount of information, but in a bit more native format. Future changes on top to actually make `Ty: Debug` call to `TyKind: Debug` should avoid going thru any of those misleading output changes for the recursive calls to debug.

I guess I'm also not opposed to keeping the output a bit more "`derive(Debug)`-like" (like using parentheses as delimiters for info).

Comment (bjorn3) [reply compiler/rustc_type_ir/src/sty.rs:502]: You got me wrong. I mean that I prefer if `Debug for TyKind` stays the same, and instead `Debug for Ty` uses pretty printing and if you want pretty printing you can format `Ty` instead of `TyKind`.

File: compiler/rustc_middle/src/ty/structural_impls.rs (hunk 3 above)
Comment (BoxyUwU) [reply compiler/rustc_middle/src/ty/structural_impls.rs:99]: ah good idea

File: compiler/rustc_type_ir/src/sty.rs (hunk 4 above)
Comment (BoxyUwU) [reply compiler/rustc_type_ir/src/sty.rs:502]: I don't see any win for printing as `Bool` over `bool` otoh `bool` is more like the syntax everybody knows. debug printing `ty.kind()` is reasonable rn since `Ty` doesn't actually have a Debug impl. `TyKind` just does not work as a debug impl, its far too verbose to reasonably use to actually debug anything.

The only important thing for a `Debug` impl here is that it doesnt lose any important information (which this PR doesnt cause us to do). It's already too late for `TyKind::Debug` to be "just how we represent", infer vars have a debug impl that is more concise and using the notation that people use to talk about infer vars in technical convos. It makes more sense for the debug impl for how we represent types to be useful for people debugging the type system than for people who dont know how `TyKind` works.

We actually cant `{:?}` `Ty` right now either, that's just a wrapper for `with_no_trimmed_paths` around the display impl (ofc eventually we should change this to defer to `TyKind: Debug`). The goal is not for `TyKind: Debug` to be "rust syntax", the goal is for it to be an output suitable for debugging a `TyKind`. Right now that is not possible.

**Review by compiler-errors — APPROVED** [review]
this is fine for now, though notably incomplete until we get `<Ty as Debug>` to use `<TyKind as Debug>` i think
